arangod/Cluster/ClusterMethods.cpp

```diff
@@ -203,29 +203,27 @@ OperationResult handleResponsesFromAllShards(
   // If none of the shards responds we return a SERVER_ERROR;
   Result result;
   VPackBuilder builder;
+
   pre(result, builder);
-  if (result.fail()) {
-    return OperationResult(result, builder.steal());
-  }
-  for (Try<arangodb::network::Response> const& tryRes : responses) {
-    network::Response const& res = tryRes.get();  // throws exceptions upwards
-    ShardID sId = res.destinationShard();
-    int commError = network::fuerteToArangoErrorCode(res);
-    if (commError != TRI_ERROR_NO_ERROR) {
```

Comment:
so this is really a waste of everones time. if `res.error == fuerte::Error::NoError` then `res.response != nullptr`

Reply:
Except that it crashed there

Reply:
thats a bug then which is nicely covered up now

Reply:
I can't see any guarantee in the code that `response` is not a nullptr if `error` is `NoError`. 
Creating a new netowrk::Response object has a nullptr `response` value, and an undefined `error` value.

Reply:
In fact, the existing code suggests this is **not** always true, for example:
```
 54   // returns a slice of the payload if there was no error
 55   [[nodiscard]] velocypack::Slice slice() const {
 56     if (error == fuerte::Error::NoError && response) {
 57       return response->slice();
 58     }
 59     return velocypack::Slice();  // none slice
 60   } 
```
Why would someone check if `response` is set here if it were guaranteed that in case of `error == NoError` it is set?

Reply:
i guess it was a precaution. however getting NoError withou a response does not make much sense

Reply:
That may be the desired behavior. But it's not guaranteed right now.
Just look at this, for example:
```
1302 struct AsyncRequest {
1303   network::Response response;
1304   std::string destination;
1305   CoordTransactionID coordTransactionID;
1306   OperationID operationID;
1307   bool done = false;
1308 };

...
1553   auto ar = std::make_shared<::AsyncRequest>();
1554   ar->destination = destination;
1555   ar->operationID = opId;
1556   ar->coordTransactionID = coordTransactionID;
1557 
1558   network::sendRequest(pool, destination, reqType, path, std::move(body), reqOpts)
1559   .thenValue([ar] (network::Response&& r) {
1560     {
1561       std::lock_guard<std::mutex> guard(::_requestMutex);
1562       ar->response = std::move(r);
1563       ar->done = true;
1564     }
1565     ::_requestCV.notify_all();
1566   });
```

Reply:
so what thats different code

Reply:
Ìt's from arangod/Cluster/v8-cluster.cpp
It defines a struct AsyncRequest with a default-constructed `network::Response` member.
As network::Response doesn't have any user defined constructors, its members will be default-initialized.
`error` is a uint16_t, so it will have an arbitrary value, while `response` will be properly constructed as a unique_ptr pointing to 0.